sklearn/utils/_array_api.py

```diff
@@ -552,16 +692,48 @@ def get_namespace(*arrays, remove_none=True, remove_types=(str,), xp=None):
     return namespace, is_array_api_compliant
 
 
-def get_namespace_and_device(*array_list, remove_none=True, remove_types=(str,)):
-    """Combination into one single function of `get_namespace` and `device`."""
+def get_namespace_and_device(
+    *array_list, remove_none=True, remove_types=(str,), xp=None
```

Comment:
I kept the `xp` argument around for `get_namespace_and_device` since it just delegates to `get_namespace`, but I agree that it's odd that `get_namespace` allows a passthrough... The only functionality of `get_namespace(xp=...)` would be to check if array api compatibility is enabled in the config, but it would be weird then to have an array namespace to pass in...

In short, I think removing it from the `get_namespace*` methods makes sense, or changing the behavior to verify that the provided namespace matches the namespace of the input arrays. If it is removed, callers would just have to gate the `get_namespace` call behind a `xp is None` check.

A quick perusal through the codebase suggests the `xp` kwarg is used in:
- `linear_model._ridge._solve_svd()`
- `metrics._regression._check_reg_targets()`
- `utils._array_api`
  - `_asarray_with_order`
  - `_expit`
  - `_nansum`
  - `_nanmax`
  - `_nanmin`
  - `_ravel`
  - `get_namespace_and_device`

Reply:
I think it's good not to have to write `if xp is not None` as much as possible.

So I am ok to pass the `xp` kwarg down to `get_namespace`. But we might also want to change `get_namespace` so that it actually check at all arrays passed to `get_namspace` actually stem from `xp` when it is not `None`.

WDYT?

Maybe this change is better be done in a dedicated side PR as it seems a bit unrelated to the purpose of this PR.

Reply:
Ok I re-read @charlesjhill's reply and I changed my mind again. I think it's good to keep the skip through behavior (without additional checks) when passing `xp != None` as there are many private helper methods that call `get_namespace*` and that accept an optional `xp`. Sometimes they are called by a public method that did not do the namespace inspection and rely on the private helper to do it, other times they are called by methods that already did the inspection upper in their callers ancestry.

The alternative would be to make the `xp` parameter mandatory in all private helper methods and always call `get_namespace*` in the first public API method and nowhere else.

If we decide for the latter, I think we should do this refactoring in a dedicated PR.

Reply:
So, to summarize, the consensus is to keep `xp` as an optional kwarg for the `get_namespace*` methods, and when it is provided, to not perform any validation (i.e., current behavior)?